### PR TITLE
2D Registration Pipeline Improvements

### PR DESCRIPTION
**Pull request by csparker247**
- Make automatic landmark detection work
- Add support for 8bpc, RGBA images

### CODE REVIEW

File: apps/src/Register2D.cpp [hunk 1]
```diff
@@ -91,16 +99,40 @@ int main(int argc, char* argv[])
...
-        fs::path landmarksFileName = parsed["landmarks"].as<std::string>();
-        LandmarkReader landmarkReader(landmarksFileName);
-        landmarkReader.setFixedImage(fixedImage);
-        landmarkReader.setMovingImage(movingImage);
-        landmarkReader.read();
-        auto fixedLandmarks = landmarkReader.getFixedLandmarks();
-        auto movingLandmarks = landmarkReader.getMovingLandmarks();
+    if (parsed.count("disable-landmark") == 0) {
+        LandmarkContainer fixedLandmarks, movingLandmarks;
+        if (parsed.count("input-landmarks") > 0) {
```
Comment (stephenrparsons) [inline apps/src/Register2D.cpp:104]:
```suggestion
        if (parsed.count("input-landmarks")) {
```

Comment (stephenrparsons) [reply apps/src/Register2D.cpp:104]: Does this do the same thing?

File: rtlib/src/LandmarkDetector.cpp [hunk 2]
```diff
@@ -20,46 +20,78 @@ std::vector<rt::LandmarkPair> LandmarkDetector::compute(int numMatches)
...
+    featureDetector->detectAndCompute(
         movingImg_, movingMask_, movingKeyPts, movingDesc, false);
 
     // Compute matches from descriptors
-    auto matcher = cv::DescriptorMatcher::create("BruteForce-Hamming");
-    std::vector<cv::DMatch> matches;
-    matcher->match(fixedDesc, movingDesc, matches, cv::Mat());
+    cv::BFMatcher matcher(cv::NORM_HAMMING);
+    std::vector<std::vector<cv::DMatch>> matches;
+    matcher.knnMatch(fixedDesc, movingDesc, matches, 2);
```
Comment (stephenrparsons) [inline rtlib/src/LandmarkDetector.cpp]:
```suggestion
    int k = 2;
    matcher.knnMatch(fixedDesc, movingDesc, matches, k);
```

File: rtlib/src/LandmarkDetector.cpp [hunk 3]
```diff
@@ -20,46 +20,78 @@ std::vector<rt::LandmarkPair> LandmarkDetector::compute(int numMatches)
...
-        auto movIdx = m.trainIdx;
-        output_.emplace_back(fixedKeyPts[fixIdx].pt, movingKeyPts[movIdx].pt);
+            goodFixed.emplace_back(fixedKeyPts[fixIdx].pt);
+            goodMoving.emplace_back(movingKeyPts[movIdx].pt);
+        }
+    }
+
+    // RANSAC outlier filtering
+    cv::Mat ransacMask;
+    cv::findHomography(goodMoving, goodFixed, cv::RANSAC, 3, ransacMask);
```
Comment (stephenrparsons) [inline rtlib/src/LandmarkDetector.cpp]:
```suggestion
    double ransacReprojThreshold = 3;
    cv::findHomography(goodMoving, goodFixed, cv::RANSAC, ransacReprojThreshold, ransacMask);
```

**Review by stephenrparsons — APPROVED** [review]
Looks good to me and I am excited to use it

File: apps/src/Register2D.cpp (hunk 1 above)
Comment (csparker247) [reply apps/src/Register2D.cpp:104]: It does. I got in the habit of avoiding it because implicit conversion from int to bool might mask a bug.

**Pull request merged**